Source/Core/Common/DirectIOFile.cpp

```diff
@@ -0,0 +1,250 @@
+// Copyright 2025 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "Common/DirectIOFile.h"
+
+#include <cassert>
+#include <utility>
+
+#if defined(_WIN32)
+#include <string_view>
+
+#include <windows.h>
+
+#include "Common/CommonFuncs.h"
+#include "Common/StringUtil.h"
+#else
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#endif
+
+#include "Common/Assert.h"
+
+namespace File
+{
+DirectIOFile::DirectIOFile() = default;
+
+DirectIOFile::~DirectIOFile()
+{
+  Close();
+}
+
+DirectIOFile::DirectIOFile(const DirectIOFile& other)
+{
+  *this = other.Duplicate();
+}
+
+DirectIOFile& DirectIOFile::operator=(const DirectIOFile& other)
+{
+  return *this = other.Duplicate();
+}
+
+DirectIOFile::DirectIOFile(DirectIOFile&& other)
+{
+  Swap(other);
+}
+
+DirectIOFile& DirectIOFile::operator=(DirectIOFile&& other)
+{
+  Close();
+  Swap(other);
+  return *this;
+}
+
+DirectIOFile::DirectIOFile(const std::string& path, const char open_mode[])
+{
+  Open(path, open_mode);
+}
+
+bool DirectIOFile::Open(const std::string& path, const char open_mode[])
```

Comment:
The use of a char array for the open mode implies that this supports all the "usual" flags like `b` and so on, yet it only supports `r` and `w`. I recommend using an enum instead to make it clear for callers which modes are supported.